### PR TITLE
Rewrite execution engine

### PR DESCRIPTION
**Pull request by fickludd**
The old ExecutionEngine implementation was a mess which had
accumulated over several years, and was very hard to understand.

The new ExecutionEngine is much smaller, and uses several helper
classes to handle specific aspects of query compilation.

  SchemaHelper
  Ensures that the query cache is clear when the
  schema changes, and also lock labels to allow for query execution
  without indexes being removed.

  PlanStalenessCaller
  Determines whether old plans are stale. This happens when the database
  statistics have changes enough and enough time has passed.

  CypherConfiguration
  Parses the Neo4j Config to assemble all configuration parameters
  required by the execution engine and cypher compilers.

  PreParser
  Pre-parses queries and asserts that the provided pre-parser option
  combination is valid.

Merge https://github.com/neo4j/neo4j-documentation/pull/215 to fix docs-build.

### CODE REVIEW

**Comment by fickludd** [comment]
@systay rebased and re-targeted, for your reviewing pleasure :)

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/Compatibility.scala [hunk 1]
```diff
@@ -159,27 +168,30 @@ extends LatestRuntimeVariablePlannerCompatibility[CONTEXT3_4, T, StatementV3_3](
...
-            result.maybeExecutionPlan.get
-          }
+        def createPlan(): ExecutionPlan_v3_5 = {
+          val logicalPlanStateV3_3 = compiler.planPreparedQuery(preparedQuery, contextV3_3)
+          val logicalPlanStatev3_5 = helpers.as3_4(logicalPlanStateV3_3)
+          LogicalPlanNotifications
+            .checkForNotifications(logicalPlanStatev3_5.maybeLogicalPlan.get, planContextv3_5, configv3_5)
+            .foreach(notificationLoggerv3_5.log)
+
+          // Here we switch from 3.3 to 3.4
```
Comment (pontusmelke) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/Compatibility.scala]: comment out-of-date

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/PlanStalenessCaller.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      val secondsSinceReplan = ((currentTimeMillis - f.creationTimeMillis) / 1000).toInt
+      Stale(secondsSinceReplan)
+    } else
+      NotStale
+  }
+
+  private def check(test: => Boolean, ifFalse: () => Unit ) = if (test) { true } else { ifFalse() ; false }
+}
+
+sealed trait ReusabilityInfo
```
Comment (pontusmelke) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/PlanStalenessCaller.scala]: I don't like the `Info`, maybe `ReusabilityState` or `PlanCacheState`?

**Comment by pontusmelke** [comment]
I think this looks like nice improvements. @systay would you care to take a look as well.

**Pull request merged**